### PR TITLE
GC fundamentals clean up

### LINKED ISSUE #12783: To improve performance ,disable concurrent GC
In document [ fundamentals of garbage collection ](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals )

> To improve performance when several processes are running, disable concurrent garbage collection

I think concurrent GC is help to improve application performance.It reduce the pauses time when GC is running. Why disable concurrent garbage can improve performance?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d8948cf5-0aec-308c-8046-965cc15763b2
* Version Independent ID: a30a0be5-b741-2a9c-e997-55d6d9960ca9
* Content: [Fundamentals of garbage collection](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals)
* Content Source: [docs/standard/garbage-collection/fundamentals.md](https://github.com/dotnet/docs/blob/master/docs/standard/garbage-collection/fundamentals.md)
* Product: **dotnet**
* Technology: **dotnet-standard**
* GitHub Login: @rpetrusha
* Microsoft Alias: **ronpet**

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #12783

- Add a note that concurrent GC is for 3.5 and earlier and move this section down to bottom
- Remove back to top anchors
- Style improvements

### CODE REVIEW

**Comment by mairaw** [comment]
There are some merge conflicts here

File: docs/standard/garbage-collection/fundamentals.md
Comment (NextTurn) [inline docs/standard/garbage-collection/fundamentals.md]:
```suggestion
To reserve memory, the garbage collector calls the Windows [VirtualAlloc](/windows/desktop/api/memoryapi/nf-memoryapi-virtualalloc) function and reserves one segment of memory at a time for managed applications. The garbage collector also reserves segments, as needed, and releases segments back to the operating system (after clearing them of any objects) by calling the Windows [VirtualFree](/windows/desktop/api/memoryapi/nf-memoryapi-virtualfree) function.
```

Same as line 39.

File: docs/standard/garbage-collection/fundamentals.md
Comment (gewarren) [inline docs/standard/garbage-collection/fundamentals.md]: I moved this section to the end.

File: docs/standard/garbage-collection/fundamentals.md
Comment (gewarren) [inline docs/standard/garbage-collection/fundamentals.md]: Moved this section to the end.

**Comment by mairaw** [comment]
There are still merge conflicts here. Do you want to take care of that before we start reviewing?

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >without having to free memory. [](start = 42, length = 30)

I think the original wording is better. it emphasizes that you don't need to do this manually. you cannot not free memory - it's just a matter of whether you do it yourself or GC does it for you.

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >property to compact the large object heap on demand [](start = 306, length = 51)

in core 3.0, LOH is automatically compacted when running with a hardlimit either by specifying a memory limit on a container or the configs. explained in [https://github.com/dotnet/coreclr/pull/22180](https://github.com/dotnet/coreclr/pull/22180)

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >server garbage collection can be concurrent or non-concurrent [](start = 223, length = 61)

Server GC was never concurrent before 4.5. concurrent Server GC (ie, Background Server GC) was only enabled in 4.5

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >Threads that are running native code are not suspended. [](start = 2, length = 55)

this is the same for Workstation and Server GC..this is about suspension, not the GC.

File: docs/standard/garbage-collection/fundamentals.md
Comment (BillWagner) [inline docs/standard/garbage-collection/fundamentals.md]: @Maoni0 makes a good point. But I don't like the "objects you create" phrase. It includes objects allocated by the .NET libraries. How about this wording:

```suggestion
- Enables you to develop your application without having to manually free memory.
```

File: docs/standard/garbage-collection/fundamentals.md
Comment (BillWagner) [inline docs/standard/garbage-collection/fundamentals.md]: One question I had while reading this article (Hopefully @Maoni0 can answer):  Should I assume that mobile apps built with Xamarin use the Workstation GC? We should state that explicitly.

**Comment by BillWagner** [comment]
@gewarren This really looks good. I reviewed it today, and I like it. I'd be ready to approve it, once you and @Maoni0 are happy with the content.

Let me know when you are ready for a final approval.

File: docs/standard/garbage-collection/fundamentals.md
Comment (gewarren) [inline docs/standard/garbage-collection/fundamentals.md]: @Maoni0 Please verify statement about mobile apps and workstation gc.

File: docs/standard/garbage-collection/fundamentals.md
Comment (gewarren) [reply docs/standard/garbage-collection/fundamentals.md]: See line 172.

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >If you specify server garbage collection on such a computer, the CLR uses workstation garbage collection with concurrency disabled [](start = 284, length = 130)

this is not true anymore (has never been true on core). we used to do this because there was no concurrent for Server GC.

**Comment by Maoni0** [comment]
> - A heap and a dedicated thread to perform garbage collection are provided for each CPU, and the heaps are collected at the same time. Each heap contains a small object heap and a large object heap, and all heaps can be accessed by user code. Objects on different heaps can refer to each other.

probably a good idea to introduce the SOH/LOH acronyms here

---
Refers to: docs/standard/garbage-collection/fundamentals.md:197 in 3251442. [](commit_id = 325144209fe3e41442918f8cf78282a1ba902d61, deletion_comment = False)

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >In a high memory-load situation, if all the processes start doing garbage collection, the garbage collector has 48 threads to schedule. [](start = 241, length = 135)

I would say "if all processes happen to be doing GC at the same time they would interfere with each other as there'd be 12 threads scheduled on the same processor. If these processes are active, it wouldnt be a good idea to have them all use server GC"

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >Background garbage collection [](start = 3, length = 29)

Should say "Background Workstation GC". "Background GC" is ambiguous.

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >all managed threads are suspended. [](start = 161, length = 34)

all managed threads are suspended, including the background GC thread which is a managed thread.

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: > Configure garbage collection [](start = 2, length = 29)

should this just point to the page that you are working on, @gewarren?

File: docs/standard/garbage-collection/fundamentals.md
Comment (Maoni0) [reply docs/standard/garbage-collection/fundamentals.md]: I don't think we should mention Xamarin here because it's a different runtime. 

I would say something like "Workstation garbage collection is designed for client apps and is the default of the GC flavor for stand alone apps. for hosted apps (like by asp.net) the default is decided by the host". 

I'm not sure what "This is the default setting for the gcServer element in the runtime configuration schema." means - should probably just delete it or make it more explicit by say "this means the default value for gcServer in the config is false"

File: docs/standard/garbage-collection/fundamentals.md
Comment (gewarren) [reply docs/standard/garbage-collection/fundamentals.md]: Now just deleted it altogether and link to new topic instead.

File: docs/standard/garbage-collection/fundamentals.md
Comment (gewarren) [reply docs/standard/garbage-collection/fundamentals.md]: Yes, good idea. I added the link, which will be broken until that PR is merged.

**Comment by mairaw** [comment]
I've added a label to indicate this PR is waiting on the other one first.

**Comment by gewarren** [comment]
Close/reopen for new build.

**Comment by gewarren** [comment]
@BillWagner I'm ready for a final approval. Thanks much!

**Review by BillWagner — APPROVED** [review]
This is really well organized, and a great article.

You can :shipit: when ready.

**Comment by gewarren** [comment]
@BillWagner Thanks very much!

**Pull request merged**
